internal/application/services.go

```diff
@@ -0,0 +1,228 @@
+package application
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/epinio/epinio/helpers/kubernetes"
+	"github.com/epinio/epinio/pkg/api/core/v1/models"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/util/retry"
+)
+
+// ServicesBoundApps is an extension of ServicesBoundAppsNames after it, to retrieve a map from
+// services to the full data of the applications bound to them. It uses ServicesBoundAppsNames
+// internally to quickly determine the applications to fetch.
+
+func ServicesBoundApps(ctx context.Context, cluster *kubernetes.Cluster, namespace string) (map[string]models.AppList, error) {
+
+	result := map[string]models.AppList{}
+
+	bindings, err := ServicesBoundAppsNames(ctx, cluster, namespace)
+	if err != nil {
+		return result, err
+	}
+
+	// Internal map of fetched applications.
+	fetched := map[string]*models.App{}
+
+	for key, appNames := range bindings {
+		serviceName, namespace := DecodeServiceKey(key)
+		for _, appName := range appNames {
+			app, ok := fetched[appName]
+			if !ok {
+				meta := models.NewAppRef(appName, namespace)
+				app := meta.App()
+				err := fetch(ctx, cluster, app)
```

Comment:
At some point we should find a way to avoid these N+1 queries. Maybe by implementing a version of `fetch` that fetches multiple apps at once, with as few requests as possible. These endpoints are not very uncommon so we may have multiple users running these commands are the same time, not to mention the Web UI polling which will generate quite some traffic even if the user takes no action. Not to be solved here, but let's start thinking about it.

Reply:
List, via label selectors, on each of the application-associated secrets, maybe. Similar to what we do for the tracking of services/configurations vs app.

Reply:
I will make a ticket.